Run build commands without declared outputs

neonichu opened the pull request:
It is possible to declare build commands without outputs and the expectation is that those would still run. Currently, that is not the case, since the only condition that makes llbuild run build commands is that there's a client for the output files (either compilation or copying of resources).

This change adds a phony output to any command that has no declared outputs and declares these phony outputs as input for any targets that are asking for the plugin to be applied. This will lead to these commands running unconditionally, fixing the current silent failure to run these.

rdar://100415491

neonichu commented:
I need to add a test for this.

neonichu commented:
@swift-ci please test

neonichu commented:
I wonder whether we should make this change conditional on the tools-version?

tomerd commented:
> I wonder whether we should make this change conditional on the tools-version?

seems like a good idea

rauhul commented:
Will this allow for `swift build --product [SomeLib|SomeExe]`?

neonichu commented:
> Will this allow for swift build --product [SomeLib|SomeExe]?

Not sure I understand the question, how is that related to plugins? 🤔

neonichu commented:
I think what I'll do is make the new behavior 5.11 only but emit a warning on all earlier tools-versions. It is probably never intentional to declare a plugin usage that doesn't actually do anything and it's good to inform people that upgrading will give them the desired behavior.

neonichu commented:
Rebased, added tests and made the fix conditional on the tools-version. On older tools-version, we emit a warning that should explain what's happening and that it can be resolved by upgrading the tools-version.

neonichu commented:
@swift-ci please test

neonichu commented:
@swift-ci please test windows

neonichu commented:
@swift-ci please test windows
